traffic_ops/traffic_ops_golang/invalidationjobs/invalidationjobs.go

```diff
@@ -0,0 +1,928 @@
+           SELECT deliveryservice.xml_id
+           FROM deliveryservice
+           WHERE deliveryservice.id=job.job_deliveryservice
+          ) AS deliveryservice,
+          job.id,
+          job.keyword,
+          job.parameters,
+          job.start_time
+`
+
+type apiResponse struct {
+	Alerts   []tc.Alert         `json:"alerts,omitempty"`
+	Response tc.InvalidationJob `json:"response,omitempty"`
+}
+
+// Used by GET requests to `/jobs`, simply returns a filtered list of
+// content invalidation jobs according to the provided query parameters.
+func (job *InvalidationJob) Read() ([]interface{}, error, error, int) {
+	queryParamsToSQLCols := map[string]dbhelpers.WhereColumnInfo{
+		"id":              dbhelpers.WhereColumnInfo{"job.id", api.IsInt},
+		"keyword":         dbhelpers.WhereColumnInfo{"job.keyword", nil},
+		"assetUrl":        dbhelpers.WhereColumnInfo{"job.asset_url", nil},
+		"userId":          dbhelpers.WhereColumnInfo{"job.job_user", api.IsInt},
+		"createdBy":       dbhelpers.WhereColumnInfo{`(SELECT tm_user.username FROM tm_user WHERE tm_user.id=job.job_user)`, nil},
+		"deliveryService": dbhelpers.WhereColumnInfo{`(SELECT deliveryservice.xml_id FROM deliveryservice WHERE deliveryservice.id=job.job_deliveryservice)`, nil},
+		"dsId":            dbhelpers.WhereColumnInfo{"job.job_deliveryservice", api.IsInt},
+	}
+
+	where, orderBy, queryValues, errs := dbhelpers.BuildWhereAndOrderBy(job.APIInfo().Params, queryParamsToSQLCols)
+	if len(errs) > 0 {
+		var b strings.Builder
+		b.WriteString("Reading jobs:")
+		for _, err := range errs {
+			b.WriteString("\n\t")
+			b.WriteString(err.Error())
+		}
+
+		return nil, nil, errors.New(b.String()), http.StatusInternalServerError
+	}
+
+	// TODO: check tenancy here
+	query := readQuery + where + orderBy
+
+	log.Debugln("generated job query: " + query)
+	log.Debugf("executing with values: %++v\n", queryValues)
+
+	returnable := []tc.InvalidationJob{}
+	rows, err := job.APIInfo().Tx.NamedQuery(query, queryValues)
+	if err != nil {
+		return nil, nil, fmt.Errorf("querying: %v", err), http.StatusInternalServerError
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		j := tc.InvalidationJob{}
+		err := rows.Scan(&j.ID,
+			&j.Keyword,
+			&j.Parameters,
+			&j.AssetURL,
+			&j.StartTime,
+			&j.CreatedBy,
+			&j.DeliveryService)
+		if err != nil {
+			return nil, nil, fmt.Errorf("parsing db response: %v", err), http.StatusInternalServerError
+		}
+
+		returnable = append(returnable, j)
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, nil, fmt.Errorf("Parsing db responses: %v", err), http.StatusInternalServerError
+	}
+
+	// This cannot be done in the scanning loop, because pq will throw an error if you try to make
+	// another query before exhausting the rows returned by an earlier query
+	filtered := []interface{}{}
+	for _, r := range returnable {
+		userErr, sysErr, errCode := IsUserAuthorizedToModifyDSXMLID(job.APIInfo(), *r.DeliveryService)
+		if sysErr != nil {
```

Comment:
I know now the function will always return a sysErr if a userErr is returned but to protect ourselves going forward I wonder if this should change to `sysErr != nil || userErr != nil`

Reply:
Nah. I can do better. That function should just return a boolean and an error. There's no reason to try to parse all that junk out in there, it's a useless mess.